### ISSUE TITLE
PatAlgos unit test failure after #24501 merge

### ISSUE THREAD

**Issue by fabiocos** [cms-sw/cmssw#26236]
The refactoring of RecoTau configurations in #24501 induces a problem in one of the unit tests in PhysicsTools/PatAlgos , that is failing DT/cgi-bin/logreader/slc7_amd64_gcc700/CMSSW_10_6_X_2019-03-22-2300/unitTestLogs/PhysicsTools/PatAlgos#/553-553)
with the error

```
----- Begin Fatal Exception 23-Mar-2019 06:18:16 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=RecoTauPiZeroProducer label='ak4PFJetsLegacyHPSPiZerosPFlow'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'stripEtaAssociationDistance'
 'stripPhiAssociationDistance'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

The problem happens in patTuple_PF2PAT_cfg.py and there is a related discussion in the review of #24501 concerning the parameter definition of RecoTauPiZeroProducer .

**Comment by fabiocos** [comment]
assign analysis,reconstruction

**Comment by perrotta** [comment]
@swozniewski @mbluj please have a look

The issue is visible by running the unit test PhysicsTools/PatAlgos/test/patTuple_PATandPF2PAT_cfg.py

Looking at the code, at the log of the unit test, and at the explicit configs created by fillDescriptions() in cfipython/slc7_amd64_gcc700/RecoTauTag/RecoTau I see the following problems in the pull request #24501 recently merged:
- The VPSet are incorrectly added in the code, and those vectors of PSets are not added in the python files produced, see e.g. cfipython/slc7_amd64_gcc700/RecoTauTag/RecoTau/ak4PFJetsLegacyHPSPiZeros_cfi.py
- In the pull request there are duplicated python configs, e.g. RecoTauTag/RecoTau/python/PFRecoTauProducer_cfi.py and the one created by fillDescriptions in cfipython/slc7_amd64_gcc700/RecoTauTag/RecoTau/PFRecoTauProducer_cfi.py, and both contain the same pfRecoTauProducer module.

I apologize for having overlooked them in the review.

Both issues represent serious errors, and they must be fixed quickly in all occurrencies in the PR. Please let us know if you don't think it is possible to submit a ug fix pull request within tomorrow (Monday): in that case, I suggest to revert #24501 and let the tau people prepare another pull request with all fixes included.

**Comment by perrotta** [comment]
About the parameters with the same name but different types reported in https://github.com/cms-sw/cmssw/pull/24501#discussion_r266226288, I would suggest to rename the PSet's with something like ```stripEtaAssociationDistanceFunc``` and ```tripPhiAssociationDistanceFunc``` (for example) and let them as optional parameters together with the double's ```stripEtaAssociationDistance``` and ```stripPhiAssociationDistance```: that would be a easy replace.

**Comment by swozniewski** [comment]
@perotta Isn't the problem that the label 'ak4PFJetsLegacyHPSPiZerosPFlow' is not contained in the fillDescriptions of the RecoTauPiZeroProducer? Could it be that it therefore compares with the first fillDescription added, which has the label 'ak4PFJetsLegacyTaNCPiZeros' and does not contain the requested parameters [https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/plugins/RecoTauPiZeroProducer.cc#L309-L339](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/plugins/RecoTauPiZeroProducer.cc#L309-L339) . Therefore, I guess the quick solution would be to add a dedicated fillDescription or to add it to one of the existing descriptions (if one of them already matches).

**Comment by perrotta** [comment]
@swozniewski 

I think I have a (not so) quick and (quite likely) dirty fix for the issue with the RecoTauPiZeroProducer in
https://github.com/perrotta/cmssw/tree/removeDuplicatePfRecoTauProducer_cfi
Feel free of importing that commit if you agree with it

However, after having applied it , the same patTuple_PATandPF2PAT_cfg.py unit test indicates another issue [1] with the configuration of RecoTauDiscriminantCutMultiplexer: if I look at what now implemented in RecoTauDiscriminantCutMultiplexer::fillDescriptions() [2] I see that it misses most of the parameters needed in the configurations. This has also to be fixed.

After that, I would continue with running the (quite useful!) patTuple_PATandPF2PAT_cfg.py unit test and test all configs created with #24501. Once that unit test successfully finishes, I would further take care of removing all duplicates (as PFRecoTauProducer_cfi.py in my commit above).

Only then such fix can be considered ready to be submitted as a bugfix pull request.

[1]
```
----- Begin Fatal Exception 25-Mar-2019 13:03:25 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=RecoTauDiscriminantCutMultiplexer label='discriminationByIsolationMVA2LoosePFlow'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'leadTrack'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

[2]
```
recoTauDiscriminantCutMultiplexer = cms.EDProducer('RecoTauDiscriminantCutMultiplexer',
  toMultiplex = cms.InputTag('fixme'),
  PFTauProducer = cms.InputTag('fixme'),
  verbosity = cms.int32(0),
  inputFileName = cms.FileInPath('RecoTauTag/RecoTau/data/emptyMVAinputFile'),
  loadMVAfromDB = cms.bool(True),
  Prediscriminants = cms.PSet(
    BooleanOperator = cms.string('and'),
    decayMode = cms.PSet(
      cut = cms.double(0),
      Producer = cms.InputTag('fixme')
    )
  ),
  mvaOutput_normalization = cms.string(''),
  key = cms.InputTag('fixme')
)
```

**Comment by perrotta** [comment]
@fabiocos : I think that the fixes needed to solve this issue are rather large.
Given the time constraint you have in building the next pre-release, I'd let you set a deadline for the tau group to provide a bugfix pull request. Otherwise, I would suggest to revert #24501 and wait for another pull request which includes it and all other fixes that will be needed.
Please let us know what do you think.

**Comment by fabiocos** [comment]
@swozniewski @perrotta as we are approaching the 10_6_0_pre3 deadline, and I would like to avoid delays if not strictly needed, I suggest the following: either by this evening a fix allowing the code to pass the PhysicsTools/PatAlgos unit tests is available, or l will revert #24501. In this way the authors will have the possibility to work on a proper fix without too much pressure, and in parallel we will avoid creating issues to other code developments (due to the failing test)

**Comment by swozniewski** [comment]
I'll try to provide the fix today. Being at a conference, I don't know how far I'll get so please don't wait and go ahead with reverting whenever you like.

**Comment by perrotta** [comment]
A fix is proposed in #26254
With it the unit test patTuple_PATandPF2PAT_cfg.py successfully run till completion

**Comment by fabiocos** [comment]
@perrotta @swozniewski #26254 has been successfully merged, so I consider this specific issue as closed. Any further possible revision of this code will be managed though a new PR I assume, and the next validation round may provide feedback about the already provided updates
